Catty-308 Implement Landscape Mode in Data Structure

Bortol15 opened the pull request:
In order to enable the landscape mode, a landscape bool has to be implemented in the data structure. The askUserForLandscapeMode will be implemented later on, for now it just returns false. The manipulation of the sensors will be implemented using the updateDeviceSensorForLandscapeMode method in SensorManager.swift
https://jira.catrob.at/browse/CATTY-308

- [x] Include the name of the Jira ticket in the PR’s title
- [ ] Verify that the Jira ticket is in the status *Ready for Development*
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s git workflow (rebase and squash your commits)
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *#catty* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

roelac11 commented on src/Catty/Helpers/Util/Util.swift:
```diff
@@ -95,4 +95,13 @@ func synchronized(lock: AnyObject, closure: () -> Void) {
+
+        //The Alert Dialog is going to be implemented here.
+        //landscapeMode is going to be changed on users decision
+        //For now it's just returning true
```
Please correct this comment to avoid confusion

roelac11 commented on src/Catty/ViewController/CatrobatTableViewController/CatrobatTableViewController.m:
```diff
@@ -179,6 +179,8 @@ - (void)addProjectAndSegueToItActionForProjectWithName:(NSString*)projectName
     self.defaultProject = [Project defaultProjectWithName:projectName projectID:nil];
     if ([self shouldPerformSegueWithIdentifier:segueToNewProjectIdentifier sender:self]) {
         [self hideLoadingView];
+        self.defaultProject.header.landscapeMode = [Util askUserForLandscapeMode];
```
Maybe comment this line out and add a ToDo to use it when the askUserForLandscapeMode function is implemented?

roelac11 commented on src/Catty/ViewController/Projects/MyProjectsViewController.m:
```diff
@@ -154,6 +154,8 @@ - (void)addProjectAndSegueToItActionForProjectWithName:(NSString*)projectName
     projectName = [Util uniqueName:projectName existingNames:[Project allProjectNames]];
     self.defaultProject = [Project defaultProjectWithName:projectName projectID:nil];
     if ([self shouldPerformSegueWithIdentifier:segueToNewProjectIdentifier sender:self]) {
+        self.defaultProject.header.landscapeMode = [Util askUserForLandscapeMode];
```
Same as above

Bortol15 commented:
Hey @roelac11   ! Thank you for the Code Review. I applied your requested changes. Due to that, I changed the DeviceSensor Protocol so that the changes caused by the landscapeMode happen in the protocols convertToStandardized function. I restored the protocols extension as it was before. I also updated the Unit Tests for the standardizedValue method to verify that, for now, the tests evaluate to the same value regardless of the landscapeMode.

roelac11 commented:
> Hey @roelac11 ! Thank you for the Code Review. I applied your requested changes. Due to that, I changed the DeviceSensor Protocol so that the changes caused by the landscapeMode happen in the protocols convertToStandardized function. I restored the protocols extension as it was before. I also updated the Unit Tests for the standardizedValue method to verify that, for now, the tests evaluate to the same value regardless of the landscapeMode.
> Add: the TODO's of course cause a Warning, but I guess that's okay..?

Hey, looks good!
I would say it's okay, but let's see what's Michael's opinion on that.

m-herold commented on src/CattyTests/PlayerEngine/Sensors/Date/DateMonthSensorTest.swift:
```diff
@@ -69,8 +69,8 @@ final class DateMonthSensorTest: XCTestCase {
-        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1))
-        XCTAssertEqual(10, sensor.convertToStandardized(rawValue: 10))
+        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1, landscapeMode: false))
+        XCTAssertEqual(10, sensor.convertToStandardized(rawValue: 10, landscapeMode: false))
```
Please also add a test for `convertToStandardized` when called with `landscapeMode` `true`.

Bortol15 replied:
Done!

m-herold commented on src/CattyTests/PlayerEngine/Sensors/Date/DateWeekdaySensorTest.swift:
```diff
@@ -90,25 +90,25 @@ final class DateWeekdaySensorTest: XCTestCase {
 
         /* test Saturday */
-        XCTAssertEqual(6, Int(sensor.convertToStandardized(rawValue: 7)))
+        XCTAssertEqual(6, Int(sensor.convertToStandardized(rawValue: 7, landscapeMode: false)))
```
Please also add a test for `convertToStandardized` when called with `landscapeMode` `true`.

m-herold commented on src/CattyTests/PlayerEngine/Sensors/Date/DateYearSensorTest.swift:
```diff
@@ -65,8 +65,8 @@ final class DateYearSensorTest: XCTestCase {
     func testConvertToStandardized() {
-        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1))
-        XCTAssertEqual(100, sensor.convertToStandardized(rawValue: 100))
+        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1, landscapeMode: false))
```
Please also add a test for `convertToStandardized` when called with `landscapeMode` `true`.

m-herold commented on src/CattyTests/PlayerEngine/Sensors/Date/TimeHourSensorTest.swift:
```diff
@@ -70,8 +70,8 @@ final class TimeHourSensorTest: XCTestCase {
-        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1))
-        XCTAssertEqual(10, sensor.convertToStandardized(rawValue: 10))
+        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1, landscapeMode: false))
+        XCTAssertEqual(10, sensor.convertToStandardized(rawValue: 10, landscapeMode: false))
```
Please also add a test for `convertToStandardized` when called with `landscapeMode` `true`.

m-herold commented on src/CattyTests/PlayerEngine/Sensors/Date/TimeMinuteSensorTest.swift:
```diff
@@ -65,8 +65,8 @@ final class TimeMinuteSensorTest: XCTestCase {
-        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1))
-        XCTAssertEqual(100, sensor.convertToStandardized(rawValue: 100))
+        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1, landscapeMode: false))
+        XCTAssertEqual(100, sensor.convertToStandardized(rawValue: 100, landscapeMode: false))
```
Please also add a test for `convertToStandardized` when called with `landscapeMode` `true`.

m-herold commented on src/CattyTests/PlayerEngine/Sensors/Date/TimeSecondSensorTest.swift:
```diff
@@ -65,8 +65,8 @@ final class TimeSecondSensorTest: XCTestCase {
-        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1))
-        XCTAssertEqual(100, sensor.convertToStandardized(rawValue: 100))
+        XCTAssertEqual(1, sensor.convertToStandardized(rawValue: 1, landscapeMode: false))
+        XCTAssertEqual(100, sensor.convertToStandardized(rawValue: 100, landscapeMode: false))
```
Please also add a test for `convertToStandardized` when called with `landscapeMode` `true`.

m-herold commented on src/Catty/PlayerEngine/Sensors/Motion/AccelerationXSensor.swift:
```diff
@@ -42,10 +42,14 @@
         rawValue * 9.8
     }
 
+    func standardizedValue(landscapeMode: Bool) -> Double {
```
I think this method can be removed.

Bortol15 replied:
Of course, thanks.

m-herold commented on src/Catty/PlayerEngine/Sensors/Phiro/PhiroBottomRightSensor.swift:
```diff
@@ -43,7 +43,7 @@
     }
 
-    func convertToStandardized(rawValue: Double) -> Double {
+    func convertToStandardized(rawValue: Double, landscapeMode: Bool) -> Double {
```
Please add a few tests to `PhiroSensorTest` to verify the `convertToStandardized` method.

Bortol15 replied:
Done!

m-herold commented on src/Catty/PlayerEngine/Sensors/SensorManager.swift:
```diff
@@ -27,17 +27,23 @@ import CoreMotion
 
     public static var defaultValueForUndefinedSensor: Double = 0
     private static var sensorMap = [String: Sensor]() // TODO: make instance let
+    var landscapeMode: Bool = false
```
Is it necessary to change this value after initialization? If not, please change to `let`. Also change scope to `private`.

Bortol15 replied:
No it was not necessary and your suggestion is the proper way, thanks.

m-herold commented on src/Catty/PlayerEngine/Sensors/SensorManager.swift:
```diff
@@ -27,17 +27,23 @@ import CoreMotion
         sensorList.forEach { type(of: self).sensorMap[$0.tag()] = $0 }
     }
 
+    private func setLandscapeMode(landscapeMode: Bool) {
```
Please set it directly in the initializer (will be necessary when changing to `let`).

Bortol15 replied:
Done!

m-herold commented on src/Catty/Helpers/Util/Util.swift:
```diff
@@ -95,4 +95,13 @@ func synchronized(lock: AnyObject, closure: () -> Void) {
         return platformVersionWithoutPatch
     }
+
+    class func askUserForLandscapeMode() -> Bool {
```
It is probably not the best idea to implement this function in the `Util`. Since there will be user interaction you will for sure need at least one completion block, so you will need to largely refactor this method anyhow. That's why I would recommend to completely remove it for now. Maybe in a separate ticket you can come up with some ideas on how to avoid code duplication for adding a project (`MyProjectsViewController` and `CatrobatTableViewController`). Maybe a common protocol with an extension? You could then implement the user interaction in this extension directly.
